src/elm/interval.coffee

```diff
@@ -154,106 +155,231 @@ module.exports.Ends = class Ends extends Expression
     [a, b] = @execArgs ctx
     if a? and b? then a.ends(b, @precision) else null
 
+module.exports.Expand = class Expand extends Expression
+  constructor: (json) ->
+    super
+
+  exec: (ctx) ->
+    # expand(argument List<Interval<T>>, per Quantity) List<Interval<T>>
+    [intervals, per] = @execArgs ctx
+    intervals = collapseIntervals(intervals, per)
+    results = []
+    for interval in intervals
+      if !interval?
+        continue
+      # We do not support open ended intervals since result would likely be too long
+      if !interval.low? or !interval.high?
+        return null
+
+      # if we mix date and datetime, convert to datetime
+      if ((interval.low.isDate or interval.high.isDate) and 
+      (interval.low.isDateTime or interval.high.isDateTime))
+        interval = interval.copy()
+        interval.low = interval.low.getDateTime()
+        interval.high = interval.high.getDateTime()
+      
+      # Support for Date, Datetime, Time
+      if ((interval.low.isDate and interval.high.isDate) or
+      (interval.low.isDateTime and interval.high.isDateTime) or
+      (interval.low.isTime?() and interval.high.isTime?()))
+        func = @expandDTishInterval
+
+      # Support for Quantity
+      else if (interval.low.isQuantity and interval.high.isQuantity)
+        func = @expandQuantityInterval
+
+      # Support for Decimal and Integer (note Decimals with .0 (eg 1.0) are considered Integers)
+      else if (typeof interval.low is 'number' and typeof interval.high is 'number')
+        func = @expandNumericInterval
+
+      if func?
+        items = func.call(@,interval,per)
+        return null if items == null
+        pushOntoArray(results,items)
+      # If we are here, Interval type is unsupported or mismatched
+      else
+        return null
+
+    return results
+
+  expandDTishInterval: (interval, per) ->
+    per = per ? new Quantity(value: 1, unit: interval.low.getPrecision())
+    if per.unit in ['week', 'weeks']
+      per.value *= 7
+      per.unit = 'day'
+    # return null if precision not applicable (e.g. gram, or minutes for dates)
+    return null if per.unit not in interval.low.constructor.FIELDS
+    # return null if interval ends have mismatched precision
+    return null if interval.low.getPrecision() != interval.high.getPrecision()
```

Comment:
I believe this is satisfied by Bryn's email, but let me know if you think I should ask.

Reply:
Actually, based on Bryn's email, I think different precisions _are_ allowed:
> > **5. What to do with datetime intervals where the interval ends are of different precision and no per is given.**
> > Example: expand { Interval[@2018-01-01, @2018-03] }. What precision should the results have?
>
> Since the calculation begins with the lower boundary, I'd suggest that the precision of the lower boundary should provide the default granularity for the per.

Reply:
Yes you are right, thank you.

What about the following:

Case that makes sense:
expand {[@2012-01-01T12:00, @2012-01-02T12:00:00]} per day
= { [@2012-01-01T12:00, @2012-01-02T11:59] }

But what to do here?
expand {[@2012-01-01T13:00:00, @2012-01-02T12:59]} per day
= { } ?

We could do best effort and only return when we know the result for sure.

Reply:
Right... I think that the 2nd example would return an empty list -- because we don't know that there is actually a full day in that interval (due to the uncertainty).